compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionUtils.cpp

```diff
@@ -56,6 +63,118 @@ static bool canInsOperandTieWithOutsOperand(OpOperand *insOperand) {
   return llvm::any_of(linalgOp.getOutputOperands(), canTieWithOutsOperand);
 }
 
+/// Checks if a linalg op is a simple reduction of the innermost dimensions
+/// with identity map for the input.
+static bool isReductionOnInnermostDims(linalg::LinalgOp linalgOp) {
+  SmallVector<Operation *, 4> combinerOps;
+
+  auto numInputs = linalgOp.getNumInputs();
+  if (numInputs != 1 && numInputs != 2) return false;
+
+  if (linalgOp.getNumOutputs() != 1) return false;
+
+  // Check if the result dims are d0, d1, ..., which means the reduction is done
+  // in the innermost dimensions.
+  auto output = linalgOp.getOutputOperand(0);
+  auto outputIndexingMap = linalgOp.getTiedIndexingMap(output);
+  for (const auto &en : llvm::enumerate(outputIndexingMap.getResults())) {
+    auto expr = en.value();
+    if (auto dim = expr.dyn_cast<AffineDimExpr>()) {
+      if (dim.getPosition() != en.index()) return false;
+    } else {
+      return false;
+    }
+  }
+
+  if (!matchReduction(linalgOp.getRegionOutputArgs(), 0, combinerOps) ||
+      combinerOps.size() != 1) {
+    return false;
+  }
+
+  return true;
+}
+
+/// Check if the op is elementwise and the output indexing map is identity.
+static bool isSimpleElementwise(linalg::LinalgOp op) {
+  if (op.getNumLoops() != op.getNumParallelLoops()) return false;
+
+  if (!allIndexingsAreProjectedPermutation(op)) return false;
+
+  for (OpOperand *opOperand : op.getOutputOperands()) {
+    if (!op.getTiedIndexingMap(opOperand).isIdentity()) return false;
+  }
+  return linalg::hasOnlyScalarElementwiseOp(op->getRegion(0));
+}
+
+/// Check if the use of operand is a reduction -> broadcast -> elementwise.
+/// An example:
+///   s = sum(a: <16x32xf32>) -> <16xf32>
+///   d = div(a: <16x32xf32>, broadcast(s to <16x32xf32>)) -> <16x32xf32>
+static bool isReductionBroadcastElementwise(OpOperand *operand) {
+  auto producer = operand->get().getDefiningOp<linalg::LinalgOp>();
+  auto consumer = dyn_cast<linalg::LinalgOp>(operand->getOwner());
+  if (!producer || !consumer) return false;
+
+  // TODO: might be able to support dynamic shape if the shape is the same for
+  // reduction and broadcast.
+  if (producer.hasDynamicShape() || consumer.hasDynamicShape()) return false;
```

Comment:
This is here to filter out the case where the dynamic dim of the reduction is different from the dynamic dim of the elementwise, which would cause a trouble for the CUDA codegen.

Reply:
If it is specific for CUDA, could I get an idea of what the specific issue with CUDA codegen is. I dont really follow the issue here.

Reply:
Following up on this, we dont need to check the shapes themselves (actually just checking shape gives no real information, especially in the case of dynamic shapes). All the necessary constraints can be restated in terms of indexing maps. For example, here just the indexing maps being equal below should capture the constraints necessary (as I understand them)